src/compiler/scala/tools/nsc/javac/JavaParsers.scala

```diff
@@ -721,11 +737,14 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
       }
     }
 
-    def memberDecl(mods: Modifiers, parentToken: Int): List[Tree] = in.token match {
-      case CLASS | ENUM | INTERFACE | AT =>
-        typeDecl(if (definesInterface(parentToken)) mods | Flags.STATIC else mods)
-      case _ =>
-        termDecl(mods, parentToken)
+    def memberDecl(mods: Modifiers, parentToken: Int): List[Tree] = {
+      adaptRecordIdentifier()
+      in.token match {
+        case CLASS | ENUM | RECORD | INTERFACE | AT =>
+          typeDecl(if (definesInterface(parentToken)) mods | Flags.STATIC else mods)
+        case _ =>
+          termDecl(mods, parentToken)
```

Comment:
Although I don't think it really matters, but this means that we will not longer accept some Java 8 programs, eg:

```java
// Test.java
class Test {
  record r;
}

// record.java
class record { }
```

Reply:
Is this something you think we should try to accept then? Java rejects this now, since `record` is no longer a valid class name. The issue is that we need to do more lookahead (2 tokens I think) when parsing the field to try to figure out whether `record r` is the start of a field or a nested record class.

Reply:
Sorry, I meant to say "I hope we can accept this breaking change, even if it rejects some valid Java 8 programs". Using lower case class names is very unconventional.